library/core/src/main/java/com/google/android/exoplayer2/trackselection/DefaultTrackSelector.java

```diff
@@ -2535,4 +2505,76 @@ public int hashCode() {
+
+    public TextTrackScore(
+        Format format,
+        Parameters parameters,
+        int trackFormatSupport,
+        @Nullable String selectedAudioLanguage) {
+      isWithinRendererCapabilities = isSupported(trackFormatSupport, false);
+      int maskedSelectionFlags =
+          format.selectionFlags & ~parameters.disabledTextTrackSelectionFlags;
+      isDefault = (format.selectionFlags & C.SELECTION_FLAG_DEFAULT) != 0;
+      isForced = (maskedSelectionFlags & C.SELECTION_FLAG_FORCED) != 0;
+      preferredLanguageScore = getFormatLanguageScore(format, parameters.preferredTextLanguage);
+      selectedAudioLanguageScore = getFormatLanguageScore(format, selectedAudioLanguage);
+      trackHasNoLanguage = formatHasNoLanguage(format);
+      selectUndeterminedTextLanguage = parameters.selectUndeterminedTextLanguage;
+      stringDefinesNoLang = stringDefinesNoLanguage(selectedAudioLanguage);
+    }
+
+    /**
+     * Compares this score with another.
+     *
+     * @param other The other score to compare to.
+     * @return A positive integer if this score is better than the other. Zero if they are equal. A
+     *     negative integer if this score is worse than the other.
+     */
+    @Override
+    public int compareTo(@NonNull TextTrackScore other) {
+      if (this.isWithinRendererCapabilities != other.isWithinRendererCapabilities) {
+        return this.isWithinRendererCapabilities ? 1 : -1;
+      }
+      if ((this.preferredLanguageScore > 0 || (this.selectUndeterminedTextLanguage && this.trackHasNoLanguage)) ==
+          (other.preferredLanguageScore > 0 || (other.selectUndeterminedTextLanguage && other.trackHasNoLanguage))) {
+        if (this.preferredLanguageScore > 0 || (this.selectUndeterminedTextLanguage
+            && this.trackHasNoLanguage)) {
+          if (this.isDefault != other.isDefault) {
+            return this.isDefault ? 1 : -1;
+          }
+          if (this.isForced != other.isForced) {
+            // Prefer non-forced to forced if a preferred text language has been specified. Where
+            // both are provided the non-forced track will usually contain the forced subtitles as
+            // a subset.
+            return !this.isForced ? 1 : -1;
+          }
+          return (this.preferredLanguageScore > other.preferredLanguageScore) ? 1 : -1;
+        } else {
+          if (this.isDefault != other.isDefault) {
+            return this.isDefault ? 1 : -1;
+          }
+          if ((this.isForced && (this.selectedAudioLanguageScore > 0 || (this.trackHasNoLanguage && this.stringDefinesNoLang))) !=
+              (other.isForced && (other.selectedAudioLanguageScore > 0 || (other.trackHasNoLanguage && other.stringDefinesNoLang)))) {
+            return (this.isForced && (this.selectedAudioLanguageScore > 0
+                || (this.trackHasNoLanguage && this.stringDefinesNoLang))) ? 1 : -1;
+          }
+          // Track should not be selected.
+          return -1;
```

Comment:
That's interesting. We need to return 0 here to fulfill the Comparable interface (e.g. comparing the same unsupported text format with itself should always return 0).

To ensure we don't select a track for this case, we should add a "isWithinConstraints" field similar to the AudioTrackScore. This should be something like "=hasLanguageMatch || isDefault || (isForced && hasSelectedAudioLanguageMatch)". And then we can filter out TextTrackScores that are not within the constraints in selectTextTrack().